File: panel/server/model/subscription.go
package model

import (
	"strings"
	"time"
)

const (
	SubTypeSingleFile = "single-file"
	SubTypeGitRepo    = "git-repo"
	SubAuthTypeSSH    = "ssh"
	SubAuthTypeToken  = "token"
)

type Subscription struct {
	ID                      uint       `gorm:"primarykey" json:"id"`
	Name                    string     `gorm:"size:128;not null" json:"name"`
	Type                    string     `gorm:"size:32;not null;default:'git-repo'" json:"type"`
	URL                     string     `gorm:"size:512;not null" json:"url"`
	Branch                  string     `gorm:"size:128;default:''" json:"branch"`
	Schedule                string     `gorm:"size:64;default:''" json:"schedule"`
	Whitelist               string     `gorm:"size:512;default:''" json:"whitelist"`
	Blacklist               string     `gorm:"size:512;default:''" json:"blacklist"`
	DependOn                string     `gorm:"size:512;default:''" json:"depend_on"`
	PreScript               string     `gorm:"type:text;default:''" json:"pre_script"`
	HookScript              string     `gorm:"type:text;default:''" json:"hook_script"`
	AutoAddTask             bool       `gorm:"default:false" json:"auto_add_task"`
	AutoDelTask             bool       `gorm:"default:false" json:"auto_del_task"`
	Enabled                 bool       `gorm:"default:true" json:"enabled"`
	Status                  int        `gorm:"default:0" json:"status"`
	LastPullAt              *time.Time `json:"last_pull_at"`
	SaveDir                 string     `gorm:"size:512;default:''" json:"save_dir"`
	SSHKeyID                *uint      `json:"ssh_key_id"`
	AuthType                string     `gorm:"size:16;default:''" json:"auth_type"`
	AuthUsername            string     `gorm:"size:128;default:''" json:"auth_username"`
	AuthToken               string     `gorm:"type:text;default:''" json:"-"`
	AuthTokenSecretProvider string     `gorm:"size:64;default:''" json:"-"`
	AuthTokenSecretCipher   []byte     `gorm:"type:blob" json:"-"`
	SSHKeySecretProvider    string     `gorm:"size:64;default:''" json:"-"`
	SSHKeySecretCipher      []byte     `gorm:"type:blob" json:"-"`
	CACertPath              string     `gorm:"size:512;default:''" json:"ca_cert_path"`
	SubPath                 string     `gorm:"size:512;default:''" json:"sub_path"`
	Alias                   string     `gorm:"size:128;default:''" json:"alias"`
	ForceOverwrite          *bool      `gorm:"default:true" json:"force_overwrite"`
	CreatedAt               time.Time  `json:"created_at"`
	UpdatedAt               time.Time  `json:"updated_at"`
}

func (Subscription) TableName() string {
	return "subscriptions"
}

func (s *Subscription) ToDict() map[string]interface{} {
	return map[string]interface{}{
		"id":                 s.ID,
		"name":               s.Name,
		"type":               s.Type,
		"url":                s.URL,
		"branch":             s.Branch,
		"schedule":           s.Schedule,
		"whitelist":          s.Whitelist,
		"blacklist":          s.Blacklist,
		"depend_on":          s.DependOn,
		"pre_script":         s.PreScript,
		"hook_script":        s.HookScript,
		"auto_add_task":      s.AutoAddTask,
		"auto_del_task":      s.AutoDelTask,
		"enabled":            s.Enabled,
		"status":             s.Status,
		"last_pull_at":       s.LastPullAt,
		"sub_path":           s.SubPath,
		"save_dir":           s.SaveDir,
		"ssh_key_id":         s.SSHKeyID,
		"auth_type":          s.EffectiveAuthType(),
		"auth_username":      s.AuthUsername,
		"has_auth_token":     s.HasAuthToken(),
		"has_ssh_key_secret": s.HasSSHKeySecret(),
		"ca_cert_path":       s.CACertPath,
		"alias":              s.Alias,
		"force_overwrite":    s.ForceOverwrite == nil || *s.ForceOverwrite,
		"created_at":         s.CreatedAt,
		"updated_at":         s.UpdatedAt,
	}
}

func NormalizeSubscriptionAuthType(value string) string {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "":
		return ""
	case SubAuthTypeSSH:
		return SubAuthTypeSSH
	case SubAuthTypeToken:
		return SubAuthTypeToken
	default:
		return ""
	}
}

func (s *Subscription) HasAuthToken() bool {
	if s == nil {
		return false
	}
	return strings.TrimSpace(s.AuthToken) != "" || s.HasAuthTokenSecret()
}

func (s *Subscription) HasAuthTokenSecret() bool {
	if s == nil {
		return false
	}
	return strings.TrimSpace(s.AuthTokenSecretProvider) != "" && len(s.AuthTokenSecretCipher) > 0
}

func (s *Subscription) HasSSHKeySecret() bool {
	if s == nil {
		return false
	}
	return strings.TrimSpace(s.SSHKeySecretProvider) != "" && len(s.SSHKeySecretCipher) > 0
}

func (s *Subscription) EffectiveAuthType() string {
	if s == nil {
		return ""
	}
	if normalized := NormalizeSubscriptionAuthType(s.AuthType); normalized != "" {
		return normalized
	}
	if s.HasAuthToken() || s.HasAuthTokenSecret() {
		return SubAuthTypeToken
	}
	if (s.SSHKeyID != nil && *s.SSHKeyID > 0) || s.HasSSHKeySecret() {
		return SubAuthTypeSSH
	}
	return ""
}

type SubLog struct {
	ID             uint      `gorm:"primarykey" json:"id"`
	SubscriptionID uint      `gorm:"index;not null" json:"subscription_id"`
	OperationID    string    `gorm:"size:64;index;default:''" json:"operation_id"`
	Status         int       `gorm:"default:0" json:"status"`
	Content        string    `gorm:"type:text" json:"content"`
	LogCursor      int64     `gorm:"default:0" json:"log_cursor"`
	Duration       float64   `gorm:"default:0" json:"duration"`
	CreatedAt      time.Time `json:"created_at"`

	Subscription *Subscription `gorm:"foreignKey:SubscriptionID" json:"-"`
}

func (SubLog) TableName() string {
	return "sub_logs"
}

func (l *SubLog) ToDict() map[string]interface{} {
	result := map[string]interface{}{
		"id":              l.ID,
		"subscription_id": l.SubscriptionID,
		"operation_id":    l.OperationID,
		"status":          l.Status,
		"content":         l.Content,
		"log_cursor":      l.LogCursor,
		"duration":        l.Duration,
		"created_at":      l.CreatedAt,
	}
	if l.Subscription != nil {
		result["subscription_name"] = l.Subscription.Name
	}
	return result
}

File: app/lib/shared/models/subscription.dart
class Subscription {
  final int id;
  final String name;
  final String type;
  final String url;
  final String branch;
  final String? subPath;
  final String schedule;
  final String whitelist;
  final String blacklist;
  final bool autoAddTask;
  final bool autoDelTask;
  final bool enabled;
  final double status;
  final DateTime? lastPullAt;
  final String saveDir;
  final int? sshKeyId;
  final String authType;
  final String authUsername;
  final String authToken;
  final bool hasAuthToken;
  final String alias;
  final String dependOn;
  final String hookScript;
  final bool? forceOverwrite;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Subscription({
    required this.id,
    required this.name,
    this.type = 'public-repo',
    this.url = '',
    this.branch = '',
    this.subPath,
    this.schedule = '',
    this.whitelist = '',
    this.blacklist = '',
    this.autoAddTask = false,
    this.autoDelTask = false,
    this.enabled = true,
    this.status = 0,
    this.lastPullAt,
    this.saveDir = '',
    this.sshKeyId,
    this.authType = '',
    this.authUsername = '',
    this.authToken = '',
    this.hasAuthToken = false,
    this.alias = '',
    this.dependOn = '',
    this.hookScript = '',
    this.forceOverwrite,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get isRunning => status == 2;
  bool get isPulling => status == 2;
  String get normalizedType {
    switch (type) {
      case 'file':
        return 'single-file';
      case 'public-repo':
      case 'private-repo':
        return 'git-repo';
      case '':
        return 'git-repo';
      default:
        return type;
    }
  }

  bool get isSingleFile => normalizedType == 'single-file';
  bool get isGitRepo => normalizedType == 'git-repo';
  String get typeLabel {
    if (isSingleFile) return '单文件';
    if (isGitRepo) return 'Git 仓库';
    return normalizedType;
  }

  String get statusText {
    if (isRunning) return '拉取中';
    if (enabled) return '已启用';
    return '已禁用';
  }

  factory Subscription.fromJson(Map<String, dynamic> json) {
    final sshKeyId = _intOrNull(json['ssh_key_id']);
    final hasAuthToken = json['has_auth_token'] == true;
    final rawAuthType = json['auth_type']?.toString().trim() ?? '';
    final authType = rawAuthType.isNotEmpty
        ? rawAuthType
        : hasAuthToken
        ? 'token'
        : sshKeyId != null
        ? 'ssh'
        : '';
    return Subscription(
      id: _int(json['id']),
      name: json['name']?.toString() ?? '',
      type: json['type']?.toString() ?? 'public-repo',
      url: json['url']?.toString() ?? '',
      branch: json['branch']?.toString() ?? '',
      subPath: json['sub_path']?.toString(),
      schedule: json['schedule']?.toString() ?? '',
      whitelist: json['whitelist']?.toString() ?? '',
      blacklist: json['blacklist']?.toString() ?? '',
      autoAddTask: _boolOrNull(json['auto_add_task']) ?? false,
      autoDelTask: _boolOrNull(json['auto_del_task']) ?? false,
      enabled: _boolOrNull(json['enabled']) ?? true,
      status: _double(json['status']),
      lastPullAt: _date(json['last_pull_at']),
      saveDir: json['save_dir']?.toString() ?? '',
      sshKeyId: sshKeyId,
      authType: authType,
      authUsername: json['auth_username']?.toString() ?? '',
      authToken: json['auth_token']?.toString() ?? '',
      hasAuthToken: hasAuthToken,
      alias: json['alias']?.toString() ?? '',
      dependOn: json['depend_on']?.toString() ?? '',
      hookScript: json['hook_script']?.toString() ?? '',
      forceOverwrite: _boolOrNull(json['force_overwrite']),
      createdAt: _date(json['created_at']) ?? DateTime.now(),
      updatedAt: _date(json['updated_at']) ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
    'name': name,
    'type': normalizedType,
    'url': url,
    'branch': branch,
    'sub_path': subPath ?? '',
    'schedule': schedule,
    'whitelist': whitelist,
    'blacklist': blacklist,
    'auto_add_task': autoAddTask,
    'auto_del_task': autoDelTask,
    'save_dir': saveDir,
    'ssh_key_id': authType == 'ssh' ? sshKeyId : null,
    'auth_type': authType,
    'auth_username': authType == 'token' ? authUsername : '',
    if (authType == 'token' && authToken.isNotEmpty) 'auth_token': authToken,
    'alias': alias,
    'depend_on': dependOn,
    'hook_script': hookScript,
    'force_overwrite': forceOverwrite ?? true,
  };
}

int _int(dynamic v) => _intOrNull(v) ?? 0;
int? _intOrNull(dynamic v) => v is num
    ? v.toInt()
    : int.tryParse(v?.toString().trim() ?? '');
double _double(dynamic v) => v is num
    ? v.toDouble()
    : double.tryParse(v?.toString().trim() ?? '') ?? 0.0;
bool? _boolOrNull(dynamic value) {
  if (value is bool) return value;
  if (value is num) return value != 0;
  if (value is String) {
    final normalized = value.trim().toLowerCase();
    if (normalized == 'true' || normalized == '1') return true;
    if (normalized == 'false' || normalized == '0') return false;
  }
  return null;
}

DateTime? _date(dynamic v) {
  if (v is String && v.isNotEmpty) return DateTime.tryParse(v);
  return null;
}
